financial-templates-lib/price-feed/utils.js

```diff
@@ -0,0 +1,130 @@
+const lodash = require("lodash");
+const assert = require("assert");
+
+// Downloads blocks and caches them for certain time into the past.
+// Allows some in memory searches to go from timestamp to block number.
+exports.BlockHistory = (web3, blocks = []) => {
```

Comment:
i've kind of converged on this pattern over time. i use this in place of a class which maintains state. there are several benefits over classes:
- the biggest benefit is that classes have no async construction, a function like this can
- it has true private variables
- avoids using `this` and therefore eliminates an entire class of potential bugs. 
- uses composition over inheritance. you can combine interfaces through object destructuring

Reply:
I think the first bullet is pretty nice. For classes, you end up having to build factory functions to accomplish this (that's effectively how the price feeds are constructed).

The second is nice and kind of covers up (what I think is) a flaw in the js class system.

3 seems reasonable, although IMO, closure scoping is _harder_ to reason about since it's completely implicit (there's no `this` to tell you the lifetime/expected scope of the variable).

4 seems less flexible since classes can inherit _and_ compose.

Overall, I think it's totally reasonable to use this syntax here -- I think it covers up for some of the parts of js's class syntax that are lacking.

I think one of the things that makes me slightly uncomfortable (and has more to do with the way javascript manages variable lifetimes) is that closures magically extend the lifetime of any variable in their declaration environment (not sure if it's even required for them to reference the variable, although it might be). My C++ brain has this nagging feeling that the variable is homeless and has no stack or scope that's managing its lifetime or holding onto it :). Anyway, it's a super interesting way of grouping data and functionality!

Reply:
- another benefit is that you dont need to use `new` with closures

3: im not sure how `this` makes you more confident in the lifetime of the memory. i just assume that the object closure encapsulates all that memory, so when the object goes out of scope everything gets cleaned up. i guess i never really thought about it that much, but i also have probably built up a lot of intuitions over time that make it seem normal to me.

4: JS only really has as an illusion of classes, or syntactic sugar. Im not advocating that classes in general are bad or are superseded by composition ( though maybe theres a case for it), I just mean in the context of JS they dont come with any particular upsides that compel me, and significant downsides that i want to avoid.

to your last point, i just try things that seem good, and if nothing breaks i keep doing it. I actually haven't thought that deeply about the underlying mechanics of the runtime. So far i haven't seen any memory issues using this pattern, though i don't rule out there might be subtle problems with it!

Reply:
High-level comment: I think everything you're doing here is 👍 . Everything below is just opining on different mental models for state :).

> 3: im not sure how `this` makes you more confident in the lifetime of the memory. i just assume that the object closure encapsulates all that memory, so when the object goes out of scope everything gets cleaned up. i guess i never really thought about it that much, but i also have probably built up a lot of intuitions over time that make it seem normal to me.

In my mind, it's less about memory and more about how I think about ownership. It's easiest to reason about ownership as singular. This means that state is owned by either an active stack or another piece of encapsulating state. That state can be referenced and modified elsewhere, but it has one true owner who dictates where it gets passed, how it's used, etc. Every other piece of code effectively references it in an ephemeral way.

Shared ownership is sometimes necessary and helpful, but I find it to be more confusing when trying to reason about correctness since it's not always clear where else this object may be referenced and how it may be being modified or used.

With classes, the class is clearly the owner. The state doesn't exist without the class (conceptually).

With this closure class style, each function basically has shared ownership of the closure state. At the end of the day, they end up working the same way, but it's that ownership that I find sometimes difficult to wrap my mind around since the functions can be split apart by simple deconstruction and carry the shared state ownership with them.

I mention C++ in my previous comment because my way of thinking on this is very much rooted in the way that C++ manages object ownership and lifetimes. This ownership is extremely explicit with the unique and shared pointer primitives. This is definitely a _me_ thing, btw. And a lot of it comes out of explicit memory management. Javascript is garbage collected and built to avoid these constraints, so it gives the developer a lot more flexibility over how they want to think about ownership. Because of this, there are no perf problems with doing things this way.

> 4: JS only really has as an illusion of classes, or syntactic sugar. Im not advocating that classes in general are bad or are superseded by composition ( though maybe theres a case for it), I just mean in the context of JS they dont come with any particular upsides that compel me, and significant downsides that i want to avoid.

Yeah, I get that. I mostly like the class as a conceptual model for grouping state and functionality. In JS, the construct isn't implemented that well and has quite a few strange constraints.

> to your last point, i just try things that seem good, and if nothing breaks i keep doing it. I actually haven't thought that deeply about the underlying mechanics of the runtime. So far i haven't seen any memory issues using this pattern, though i don't rule out there might be subtle problems with it!

Yeah, definitely no actual memory problems. It's a 100% sensible and performant pattern!